fix: uniqueness constraint on Tagged Items

alexkim205 opened the pull request:
## Changes

This PR fixes the uniqueness constraint on Tagged Items. Today's constraint (`unique_together`) didn't throw an IntegrityError when duplicate `TaggedItem`'s were added to the table. This happens because PG doesn't respect [uniqueness on null columns](https://stackoverflow.com/questions/33307892/django-unique-together-with-nullable-foreignkey). The solution is to explicitly construct every permutation of a Uniqueness constraint that accounts for null conditions.

This is a prerequisite to getting `bulk_create(ignore_conflicts=True)` to work for the rewrite of the failed tags migration.

Replaces the `unique_together` Meta field with UniqueConstraint since the former is on the [deprecation](https://docs.djangoproject.com/en/4.0/ref/models/options/#unique-together) path.

Note: We don't have to worry about any existing tagged items violating this uniqueness rule, since the TaggedItemSerializer already enforces this uniqueness, just not at the database level yet. 

*Stay up-to-date with our [coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

Ran a separate test migration and tested locally to check that IntegrityErrors are being thrown on non-unique inserts.

alexkim205 committed:
fix uniqueness

alexkim205 committed:
took out base by accident

alexkim205 committed:
revert base

alexkim205 committed:
black

rcmarron commented on posthog/models/tagged_item.py:
```diff
@@ -2,23 +2,42 @@
+# uniqueness across null columns.
+def build_uniqueness_constraint():
+    built_check_list: List[UniqueConstraint] = [
+        UniqueConstraint(fields=("tag",) + RELATED_OBJECTS, name=f"unique_tagged_item"),
```
What does this constraint do? I understand the others, but not sure I get this one.

Super nit: the name string doesn't have to be a f-string

alexkim205 replied:
Ah good catch, this is a case that the exactly one related object check already accounts for.

rcmarron reviewed (approved):
One comment, but otherwise it looks good to me.

Before merging, it might make sense to get @hazzadous or @guidoiaquinti to take a quick look. It seems super related to https://github.com/PostHog/posthog/pull/8825

alexkim205 committed:
remove check

hazzadous commented on posthog/models/tagged_item.py:
```diff
@@ -56,8 +73,10 @@ class TaggedItem(UUIDModel):
-        unique_together = ("tag",) + RELATED_OBJECTS
-        constraints = [models.CheckConstraint(check=build_check(), name="exactly_one_related_object",)]
+        constraints = [
+            *build_uniqueness_constraint(),
```
non-blocking

Can we just add these constraints as, for example:

```
constraints = [
  build_partial_uniqueness_constraint(field="dashboard"),
  build_partial_uniqueness_constraint(field="insight"),
  ...
]
```

I find it quite indirect with `build_uniqueness_constraint` to see at a glance what constraints are added.

alexkim205 replied:
One thing I wanted to avoid was to add one more place to check whenever someone wanted to add an object field to this model. I could also do something like: 

```
*[build_partial_uniqueness_constraint(field=o) for o in objects]
```

which might make it known what constraints we're actually making

hazzadous replied:
What does the addition of a new object field look like at the moment?

hazzadous commented on posthog/models/tagged_item.py:
```diff
@@ -2,23 +2,40 @@
+# uniqueness across null columns.
+def build_uniqueness_constraint():
+    built_check_list: List[UniqueConstraint] = []
+    for o_field in RELATED_OBJECTS:
+        built_check_list.append(
+            UniqueConstraint(
+                fields=["tag", o_field],
+                name=f"unique_{o_field}_tagged_item",
+                condition=Q(
+                    *[(_o_field, None) for _o_field in RELATED_OBJECTS if _o_field != o_field], _connector="AND"
```
Would `condition=Q(**{f"{_o_field}__isnull": False})` be more straight forward here?

Twixes replied:
It seems like this code could be a bit clearer with just some more concrete names in place of `o_field` and `_o_field` – they confused me for a moment

hazzadous reviewed (dismissed):
Some readability comments, feel free to leave.

I would like to see a test however before merging.

It's also worth noting that these constraints will be blocking, e.g. the indexes are going to be created non-concurrently. I don't think this is an issue with the tagitem table (famous last words) as this isn't actually being used at the moment and is empty?

alexkim205 commented:
Thanks for the reviews 🙏! I've addressed all feedback and written a test which exposed some potential gotcha's. 

1. Migration fails in tests even though it passes in dev environment with this message:

```
cannot CREATE INDEX "posthog_taggeditem" because it has pending trigger events
```

It seems that the concurrency issue will be an actual problem. Explicitly moving these migrations into a `migrations.SeparateDatabaseAndState` fixes and adding CONCURRENTLY into the SQL that Django creates seems to work.

2. There is the potential that this migration fails if there are existing rows that violate this unique constraint. However, there are no duplicates in the TaggedItem table which shouldn't be an issue:

```
Select action_id, event_definition_id, property_definition_id, dashboard_id, insight_id, tag_id, count(*)
from posthog_taggeditem
group by action_id, event_definition_id, property_definition_id, dashboard_id, insight_id, tag_id
HAVING count(*) > 1
```

The above returns 0 results. I think we can also protect against this a bit stronger by adding NOT VALID to the query so we don't check existing rows that might violate this constraint.

alexkim205 committed:
create indices concurrently and add a test

alexkim205 committed:
Merge remote-tracking branch 'origin' into chore/fix-uniqueness-constraint-tagged-items

hazzadous commented on posthog/migrations/0216_uniqueness_constraint_tagged_items.py:
```diff
@@ -0,0 +1,139 @@
+        migrations.RunSQL(
+            sql="""
+                CREATE UNIQUE INDEX CONCURRENTLY "unique_dashboard_tagged_item" ON "posthog_taggeditem" ("tag_id", "dashboard_id")
+                WHERE ("insight_id" IS NULL AND "event_definition_id" IS NULL AND "property_definition_id" IS NULL AND "action_id" IS NULL);
```
Maybe I'm missing something, but what is the issue with just specifying `WHERE dashboard_id IS NOT NULL` ?

This would seem less brittle for example when adding another field.

alexkim205 replied:
You're completely right, sorry if I missed that in the previous review! I ran the tests with this changed to NOT NULL and seems to work the same.

hazzadous commented on posthog/test/test_migration_0216.py:
```diff
@@ -0,0 +1,36 @@
+        TaggedItem = self.apps.get_model("posthog", "TaggedItem")  # type: ignore
+
+        with self.assertRaises(IntegrityError):
+            TaggedItem.objects.create(dashboard_id=self.dashboard.id, tag_id=self.tag.id)
```
Given it's very low effort I would add in tests for each of the model types.

non-blocking: not for this PR, but is there a more stable API level we can test at? e.g. at the REST api level, I would add something like:

```
def test_tagging_items_are_unique():
    PUT /api/insights/abc/tags/some-tag
    PUT /api/insights/abc/tags/some-tag
    GET /api/insights/abc/tags/

    assert there is just one tag
```

It's worth noting that this isn't just a test related to this specific migration, but a requirement/feature of the API

alexkim205 replied:
Good point, that would've probably been an easier test to write than the migration test. I'll add one test for each object.

alexkim205 replied:
Also probably a good idea to add model level tests too.

hazzadous commented on posthog/models/tagged_item.py:
```diff
@@ -2,23 +2,36 @@
+        fields=["tag", field],
+        name=f"unique_{field}_tagged_item",
+        condition=Q(
+            *[(f"{other_field}__isnull", True) for other_field in RELATED_OBJECTS if other_field != field],
```
I'm still unsure why this needs to know about other fields and not just check for nullness of the field in question

hazzadous reviewed (dismissed):
I'm still unsure about the partial index condition (i.e. why do they need to depend on other fields), it looks like it will be the source of future bugs especially with adding fields.

> There is also the potential that this migration fails if there are existing rows that violate this unique constraint. However, there are no duplicates in the TaggedItem table which shouldn't be an issue:

Is this production user data? Do we need to consider self hosted data here i.e. it might be ok for cloud but what happens for self hosted? Can we just truncate as the first operation if this is just garbage data?

alexkim205 committed:
use not null

alexkim205 committed:
Merge remote-tracking branch 'origin' into chore/fix-uniqueness-constraint-tagged-items

alexkim205 committed:
dedupe as first operation

alexkim205 committed:
non deterministic tests

alexkim205 committed:
Merge remote-tracking branch 'origin' into chore/fix-uniqueness-constraint-tagged-items
